Applications/ApplicationsLib/TestDefinition.cpp

```diff
@@ -0,0 +1,215 @@
+        }
+        WARN(
+            "Calling %s from the VTKDIFF_EXE environment variable didn't work "
+            "as expected. Return value was %d.",
+            vtkdiff_exe.c_str(), return_value);
+    }
+
+    std::vector<std::string> const paths = {"", "bin"};
+    auto const path =
+        find_if(begin(paths), end(paths), [](std::string const& path) {
+            int const return_value =
+                // TODO (naumov) replace system call with output consuming call
+                // as in an above todo comment.
+                std::system((BaseLib::joinPaths(path, "vtkdiff") + " --version")
+                                .c_str());
+            return return_value == 0;
+        });
+    if (path == end(paths))
+    {
+        OGS_FATAL("vtkdiff not found.");
+    }
+    return BaseLib::joinPaths(*path, "vtkdiff");
+}
+
+}  // namespace
+
+namespace ApplicationsLib
+{
+TestDefinition::TestDefinition(BaseLib::ConfigTree const& config_tree,
+                               std::string const& reference_path,
+                               std::string const& output_directory)
+{
+    if (reference_path.empty())
+    {
+        OGS_FATAL(
+            "Reference path containting expected result files can not be "
+            "empty.");
+    }
+
+    std::string const vtkdiff = findVtkdiff();
+
+    // Construct command lines for each entry.
+    auto const& vtkdiff_configs = config_tree.getConfigSubtreeList("vtkdiff");
+    _command_lines.reserve(vtkdiff_configs.size());
+    for (auto const& vtkdiff_config : vtkdiff_configs)
+    {
+        std::string const& field_name =
+            vtkdiff_config.getConfigParameter<std::string>("field");
+        DBUG("vtkdiff will compare field '%s'.", field_name.c_str());
+
+#ifdef USE_PETSC
+        int rank;
+        MPI_Comm_rank(PETSC_COMM_WORLD, &rank);
+        int mpi_size;
+        MPI_Comm_size(PETSC_COMM_WORLD, &mpi_size);
+        std::string const& filename =
+            MeshLib::IO::getVtuFileNameForPetscOutputWithoutExtension(
+                vtkdiff_config.getConfigParameter<std::string>("file")) +
+            "_" + std::to_string(rank) + ".vtu";
+#else
+        std::string const& filename =
+            vtkdiff_config.getConfigParameter<std::string>("file");
+#endif  // OGS_USE_PETSC
+        std::string const& output_filename =
+            BaseLib::joinPaths(output_directory, filename);
+        _output_files.push_back(output_filename);
+        // TODO (naumov) expand filename relative to ref path for globbing.
+        std::string const& reference_filename =
+            BaseLib::joinPaths(reference_path, filename);
+
+        auto const& absolute_tolerance =
+            vtkdiff_config.getConfigParameterOptional<double>(
+                "absolute_tolerance");
+        std::string const absolute_tolerance_parameter =
+            absolute_tolerance == boost::none
+                ? ""
+                : "--abs " + convert_to_string(*absolute_tolerance);
```

Comment:
btw: the precedence of `+` over `:` is probably not working knowledge of everybody.